### DISCUSSION TITLE
Image opacity in stories

### DISCUSSION THREAD

**Discussion by StephanosPSteer** [pymupdf/PyMuPDF#3676] [Looking for help] [answered]
Hi 

I am using PyMUPDF to convert HTML to pdf.

When I initially started I used stories. 

Now I want to add effectively a bunch of watermarked images. 

I have tried adding the image in html 

```html

<p class="first"> <img class="myimg" src="glass_blank_bW.png"/>

``` 

with CSS

```css

.myimg {
  background-image: url("glass_blank_bW.png");
  opacity: 0.1;
  width: 100%;
}

``` 

and within code

```python

 img = ("glass_blank_bW.png", "img")  
        members = (img)  
        arch = pymupdf.Archive("/blah/blah/blah/blah/")
        arch.add(members, path="mypath")


        CSS = csspath.read_bytes().decode()
        story = pymupdf.Story(html=HTML, user_css=CSS, archive=arch)  

``` 

but I cannot get the opacity to change in the document, nor can I have the text sit on top of the image.

I can see that I can add a watermark outside of stories and that i can use insert_HTMLbox but I have no idea how to use these with stories if i need to. 

Just to add that i will want to add multiple images on different pages. 

Thanks

**Comment by JorjMcKie** [comment, marked as answer]
Yo must use separate stories or `insert_html` executions to achieve this.
A single story is intended to represent the output of one "flow" of text (actually the DOM content of the Story) across potentially multiple rectangles / pages. Although "text" may contain image references, there exists no explicit control to position details of the Story's content.
So use multiple rounds to achieve what you want. One round that writes text to pages (outputting to some memory), then re-open that new document and put your watermarks on the pages.
For round 2, `insert_htmlbox` is a convenient thing, but of course you can do whatever you want and e.g. use `insert_image` as well.

**Reply by StephanosPSteer** [reply]
Thank you that saved me so much time trying to get it to work in one go. I used insert_image after opening the document again.

**Reply by JorjMcKie** [reply]
Glad it works for you!
If you are creating a new PDF with a Story, you may not have seen an elegant way to do this:
[Story.write_with_links](https://pymupdf.readthedocs.io/en/latest/story-class.html#Story.write_with_links). It internally creates a PDF `Document` without the need to explicitly use a DocumentWriter.
That doc can then more conveniently be used to manipulate the created pages in desired ways.
